### DISCUSSION TITLE
root_path, X-Forwarded-Prefix and relative URLs

### DISCUSSION THREAD

**Discussion by kleynjan** [zauberzeug/nicegui#2813] [Q&A] [answered]
### Question

It's been a bit of a struggle to get multiple NiceGUI apps deployed on a single host name. The pattern I need is `<hostname>/<app1>/<nicegui_paths...>`, `<hostname>/<app2>/<nicegui_paths...>`, with the apps isolated from each other (separate envs, etc), each listening on another port.

APIRouter can be used to shift 'my' endpoints to a prefix, but in that case, the `/_nicegui/{version}/static` and `/_nicegui_ws/socket.io` endpoints stay attached to the root url.

The alternative is to use the `root_path="/prefix"` in uvicorn or the `proxy_set_header X-Forwarded-Prefix /prefix` in nginx (which seem to have the exact same effect). All endpoints are then available under /prefix, including the websocket. And in the documents served to the browser, the url's get automatically rewritten to include this prefix. Great!

But. Not **all** URL's get rewritten, see examples below. Hence my questions: 
1. Is there a way to tell **which** URL's are rewritten?  
2. Secondly, what is best way to add the prefix where it's missing? Add `request.scope.get("root_path")`?

Re 1: NiceGUI and/or uvicorn seem to add the prefix if it's clear that a URL is intended as part of the API, but 'url' is not a separate type and I am used to freely generating and concatenating strings which then happen to wind up as URL. AFAICT there is no mechanism "get a canonical url for endpoint X" in NiceGui?

(An alternative approach would be to always use absolute url's (including scheme and domain), but that feels like throwing out the baby with the bath water.) 

```python
from nicegui import ui, app
from fastapi import Request

# correctly mapped to /prefix/static/...!
app.add_static_files(url_path="/static", local_directory="static")

@ui.page("/")
def index(request: Request):
    print(request.scope.get("root_path"))  # -> /prefix

    ui.label("1. A path in ui.add_head_html is *not* mapped from /static/... to /prefix/static/... (see head)")
    ui.add_head_html('<link rel= "stylesheet" href="/static/styles.css">')

    ui.label("2. ui.element('img').props('src=/static/...') is *not* mapped to /prefix/static/...:")
    ui.element("img").props('src="/static/img/photo.jpg"').style("width:100px;")

    ui.html(
        "<a href='/page2'>3. The link embedded in this ui.html (<a href=....>) is *not* mapped from /page2 to /prefix/page2</a>"
    )

    ui.label(
        "4. ui.image link to '/static/img/photo.jpg' is correctly mapped to '/prefix/static/img/photo.jpg:"
    )
    ui.image("/static/img/photo.jpg").style("width: 200px;")

    ui.link("5. ui.link to /page2 is correctly mapped to /prefix/page2", "/page2")

    ui.label("6. ui.navigate.to /page2 in button is correctly mapped to /prefix/page2:")
    ui.button("Hello, world!", on_click=lambda: ui.navigate.to("/page2"))

@ui.page("/page2")
def page2():
    ui.link("ui.link to / is correctly mapped to /prefix/", "/")

## run from uvicorn with --root-path:
## edited: needs uvicorn 0.25.0 & --reload option is somehow required
#  uvicorn main:app --workers 1 --port 8080 --reload --root-path "/prefix"
ui.run()
```
<br>
Thanks in advance, 
-Peter

**Comment by kleynjan** [comment]
PS: I can't just add '/prefix' to all URLs because some would still get rewritten (eg, #4 in the example becomes /prefix/prefix/static/image/photo.jpg). 

Since pages and API endpoints can already be shifted using APIRouter, a solution would perhaps be to add a way to simply shift the `/_nicegui` and `/_nicegui_ws` endpoints? Then we wouldn't need the root_path, there would be no magical url rewriting getting in the way and proxy configs could be lot simpler.

**Comment by rodja** [comment]
@kleynjan your example is not self-contained because it needs something like nginx to do the routing. I've used our nginx_subpath example.

NiceGUI prepends the prefix where ever possible. But as you pointed out, if you have your own strings like your demos of `ui.add_head_html`, `ui.html`, `.props` etc, you need to prepend the prefix yourself. We simply compute it with

https://github.com/zauberzeug/nicegui/blob/a42c12dbf6105fd0c06fa4a37e3e9be6fb620f8f/nicegui/client.py#L116

Does it work for you if you c&p this code to your demo and add it where needed? We could make the prefix an auto-injecting variable to the page builder like `request`....

**Comment by kleynjan** [comment]
Hi @rodja, it *was* a self-contained example, but I've just found out that I was running uvicorn 0.25.0 and the handling of root_path has apparently changed between 0.25.0 and 0.29.0! I will need to see if I can get this working again (because I **need** this for a usable dev environment) and report back here.

**Comment by kleynjan** [comment]
This appears to be the issue in uvicorn: [--root-path "/" appears to break path resolution from 0.26.0 onwards #2244](https://github.com/encode/uvicorn/discussions/2244). I do see duplicate paths in 0.29.0, as is commented by @msmitherdc there. Unfortunately, it doesn't seem like a lot of people over there are feeling the pain. 

So for the time being there seems to be no way to have a dev environment that can be 100% reliably deployed to production behind a proxy. Any non-trivial app will have broken links that need to be fixed in prod (OR you'd need to run nginx in dev or separate testing env). Not ideal.

~~Would you be open to a PR that adds an optional prefix to `/_nicegui*/*`, so we can ditch this whole rewriting mess? In that case I expect to be able to deploy behind a straight nginx proxy, without the X-Forwarded-Prefix / root-path rewrites.~~

Update: just re-installed in a fresh 1.4.19 env and downgraded uvicorn to 0.25.0. Seems to work beautifully, only issue now is that I get my app served flawlessly on both / and /prefix paths. Now trying to find out how & why...

**Comment by rodja** [comment]
@kleynjan sorry, I can not follow. Are you sure https://github.com/encode/uvicorn/discussions/2244 is really your problem? I thought you are applying a subpath and have issues with some of the links. Could you make an example? The one you gave in the start of the question is explainable by having hard coded urls which do not know about the root-path / X-Fordwarded-Prefix header.

**Comment by kleynjan** [comment, marked as answer]
Yes, I've done some further work, sorry for any confusion. And you are right, uvicorn#2244 is not the issue. Let me rephrase: I want to use my NiceGUI app behind a proxy that does not strip the prefix. Tiangolo [hints at it in the docs](https://fastapi.tiangolo.com/advanced/behind-a-proxy/#about-proxies-with-a-stripped-path-prefix) but then only works out the stripped-prefix scenario.

In uvicorn 0.25.0 this happened to work (by design or by accident is unclear to me): a NiceGUI app that defines /page2 properly responds to the /prefix/page2 url, assuming the --root-path is set in uvicorn, **without** a proxy to strip the prefix. I was very happy with that, because it allows me to use the same urls in dev and prod.

In uvicorn 0.29.0 it seems (as you correctly assume) that only the stripped-prefix behavior is supported with --root-path. Hence the duplicate prefixes I see if I try to access /prefix/page2, which gets interpreted as /prefix/prefix/page2.

~~Which brings me back to my original question: I'd like to shift *all* my endpoints to /prefix, without using a proxy. APIRouter can be used to shift pages and API endpoints (eg, app responds to /prefix/page2), but I'm then still looking for a way to similarly shift /_nicegui*/* endpoints, which up to now stay attached to the root url.~~

Edit: from an architectural (ideological?) POV I'd still prefer to work with identical URLs in dev and prod (and a non-stripping proxy), but over the past days I've found I can live without it. If stripped-prefix proxying is the state of the art, I will stop trying to fight it. So feel free to file this under 'maybe someday'. :smile:

**Comment by rodja** [comment]
@kleynjan Thanks for the Edit of your last comment. I was starting to write something about the non-standard expectation to have the same path. I'll mark your comment as answer to this question.
